mllib/src/main/scala/org/apache/spark/ml/feature/Word2VecCBOWSolver.scala

```diff
@@ -0,0 +1,344 @@
+  private def generateVocab[S <: Iterable[String]](
+      input: RDD[S],
+      minCount: Int,
+      sample: Double,
+      unigramTableSize: Int): Vocabulary = {
+    val sc = input.context
+
+    val words = input.flatMap(x => x)
+
+    val sortedWordCounts = words.map(w => (w, 1L))
+      .reduceByKey(_ + _)
+      .filter{case (w, c) => c >= minCount}
+      .collect()
+      .sortWith{case ((w1, c1), (w2, c2)) => c1 > c2}
+      .zipWithIndex
+
+    val totalWordCount = sortedWordCounts.map(_._1._2).sum
+
+    val vocabMap = sortedWordCounts.map{case ((w, c), i) =>
+      w -> i
+    }.toMap
+
+    val samplingTable = new Array[Float](vocabMap.size)
+
+    if (sample > Double.MinPositiveValue) {
+      sortedWordCounts.foreach { case ((w, c), i) =>
+        val samplingRatio = sample * totalWordCount / c
+        samplingTable.update(i, (math.sqrt(samplingRatio) + samplingRatio).toFloat)
+      }
+    }
+
+    val weights = sortedWordCounts.map{ case((_, x), _) => scala.math.pow(x, power)}
+    val totalWeight = weights.sum
+
+    val normalizedCumWeights = weights.scanLeft(0.0)(_ + _).tail.map(x => x / totalWeight)
+
+    val unigramTable = generateUnigramTable(normalizedCumWeights, unigramTableSize)
+
+    Vocabulary(totalWordCount, vocabMap, unigramTable, samplingTable)
+  }
+
+  private def zeroVector(v: Array[Float]): Unit = {
+    var i = 0
+    while(i < v.length) {
+      v.update(i, 0.0f)
+      i+= 1
+    }
+  }
+
+  private def generateContextWordPairs(
+      sentence: Array[Int],
+      window: Int,
+      doSample: Boolean,
+      samplingTable: Array[Float],
+      random: XORShiftRandom): Iterator[(Array[Int], Int)] = {
+    val reducedSentence = if (doSample) {
+      sentence.filter(i => samplingTable(i) > random.nextFloat)
+    } else {
+      sentence
+    }
+    reducedSentence.iterator.zipWithIndex.map { case (word, i) =>
+      val b = window - random.nextInt(window) // (window - a) in original code
+      // pick b words around the current word index
+      val start = math.max(0, i - b) // c in original code, floor ar 0
+      val end = math.min(reducedSentence.length, i + b + 1) // cap at sentence length
+      // make sure current word is not a part of the context
+      val contextIds = reducedSentence.view.zipWithIndex.slice(start, end)
+        .filter{case (_, pos) => pos != i}.map(_._1)
+      (contextIds.toArray, word)
+    }
+  }
+
+  // This essentially helps translate from uniform distribution to a distribution
+  // resembling uni-gram frequency distribution.
+  private def generateNegativeSamples(
+      random: XORShiftRandom,
+      word: Int,
+      unigramTable: Array[Int],
+      numSamples: Int,
+      arr: Array[Int]): Unit = {
+    assert(numSamples + 1 == arr.length,
+      s"Input array should be large enough to hold ${numSamples} negative samples")
+    arr.update(0, word)
+    var i = 1
+    while (i <= numSamples) {
+      val negSample = unigramTable(random.nextInt(unigramTable.length))
```

Comment:
The unigram table by default is `100e6 * 4bytes = 400 MB` in size. The original code was meant to run on a single machine, but for Spark we are going to serialize this every iteration, which may be prohibitive. AFAICT you can just simulate the behavior needed here without actually duplicating word indices over a 100million length array. Just keep the normalized cumulative weights around and here, generate a random number between 0 and 1. Then, just do a binary search to figure out which word index it is. You only need to store O(vocabSize) numbers. The time complexity increases from O(numNonNegativeSamples) to O(numNonNegativeSamples * log2(vocabSize)), but the inner loop of the training is bound by O(numNonNegativeSamples * vectorSize), which should typically be much larger. Have you tested it at scale? Does the broadcast step take a long time? Also, what are typical vocabulary sizes? A vocab size of 1e6 is going to be about 4 MB so the space savings are pretty big.

Reply:
I agree with the analysis. I did initially implement it using the binary search method you suggested. Changing to this reduce that complexity to constant time, and I saw 5%-10% performance gain from it. I do agree about the space savings

To answer your questions:
1. I have run it at scale, on wikipedia dump and some other data sets.
2. This broadcast doesn't take too long, and happens only once at startup. Note that the weights are also collected and broadcasted every iteration, and they are also similar in size (1e6 vocab x 100 dimensions x 4bytes ~ 400Mb). 
3. Typical vocab sizes can be 500k-1e6. As a comparison [GloVe Vectors](https://nlp.stanford.edu/projects/glove/) vocab sizes range from 400k for wikipedia to 2e6 for web-crawl and such.